modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateProcessor.java

```diff
@@ -198,4 +221,35 @@ public DateProcessor create(
             );
         }
     }
+
+    private static final class Cache {
+
+        private final ConcurrentMap<Key, Function<String, ZonedDateTime>> map;
+        private final int capacity;
+
+        private Cache(int capacity) {
+            if (capacity <= 0) {
+                throw new IllegalArgumentException("cache capacity must be a value greater than 0 but was " + capacity);
+            }
+            this.capacity = capacity;
+            this.map = ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency(this.capacity);
+        }
+
+        Function<String, ZonedDateTime> getOrCompute(Key key, Supplier<Function<String, ZonedDateTime>> supplier) {
+            var element = map.get(key);
+
+            if (element != null) {
+                return element;
+            }
+
+            element = supplier.get();
+            if (map.size() >= capacity) {
+                map.clear();
```

Comment:
This is kind of surprising. Whenever the cache is full, we empty it and start over? Is this to avoid the complexity of a concurrent LRU cache? Or is the idea that we'll rarely see more than 256 formats in practice, so this is just for safety and will rarely be hit?

Reply:
Yes! this decision was made mainly because we want to avoid the complexity of LRU (or any eviction policy) within this code path. The decision to make the size 256 was primarily based on experiments with different benchmarks.

Reply:
I think it's worth adding a one line comment in the code above the clear to explain its purpose.